### PR TITLE
@pilgrimmemoirs @cheezitman

### PR DESCRIPTION
**Pull request by RedSquirrelious**
took a little longer to make it as pretty as i saw it in my head.  sure learned a lot, though.  tests are not ideal but covered it all, i do believe.

### CODE REVIEW

File: app/views/media/_buttons.html.erb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+<span class="buttons">
+	<ul class="row small-up-1 medium-up-2">
+		<li class="column">
+			<%= link_to(image_tag("/assets/upvote_cupid_dart_heart.png"), image_path(upvote_path(movie.id)), action: :upvote, method: :post) %>
```
Comment (Hamled) [inline app/views/media/_buttons.html.erb:4]: The `action: :upvote` part here is unnecessary.

Usually we only specify the `action` when we're using a hash to indicate what URL a link/button/form will go to (e.g. `{controller: 'media', action: 'upvote', id: movie.id}`).

Here we are already using `upvote_path(movie.id)` to indicate the URL and the `action: :upvote` part does not impact that at all -- it actually just becomes an HTML attribute on the `<a>` tag generated by `link_to`.

File: app/views/media/_buttons.html.erb [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+		</li>
+		<li class="column">
+			<%= link_to(image_tag("/assets/downvote_broken_heart.png"), image_path(downvote_path(movie.id)), method: :post, action: :downvote) %>
+		</li>
+		<!-- edit button -->
+		<li class="column">
+			<%= button_to("Edit", edit_path(movie.id), method: :get, action: :edit) %>
```
Comment (Hamled) [inline app/views/media/_buttons.html.erb:11]: Same deal as above with `action: :upvote`, the `action: :edit` part on here is not doing anything for us here.

File: app/views/media/_show_all_media.html.erb [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+	<% @media.each do |medium| %>
+	<article>
+		<p>Rank: <%= medium.votes.count %> </p>
+		<p>Title: <%= link_to(medium.title, show_media_path(medium.id)) %></p>
+		
+		<% if medium.kind == 'album' %>
+			<p>Artist: <%=medium.maker %></p>
+		<% elsif %>
+			<% medium.kind == 'book' %>
```
Comment (Hamled) [inline app/views/media/_show_all_media.html.erb:11]: This should go on the same line as the `elsif` like:

``` erb
<% elsif medium.kind == 'book' %>
```

File: app/views/media/show_albums.html.erb [hunk 4]
```diff
@@ -0,0 +1 @@
+<%= render partial: 'show_albums' %>
```
Comment (Hamled) [inline app/views/media/show_albums.html.erb:1]: I think if our view is just going to say to render a partial with the same name we can skip having this view altogether, and wherever we would have rendered this view, we render the partial instead.

File: app/views/media/_buttons.html.erb (hunk 1 above)
Comment (Hamled) [reply app/views/media/_buttons.html.erb:4]: Also, the use of `image_path` here is redundant. The URL we want to link to is just `upvote_path(movie.id)`, and `image_path` here doesn't modify it at all so the functionality still works.

File: test/controllers/media_controller_test.rb [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  end
+
+  test 'Downvote Method Must Destroy Votes' do
+    test_case = Medium.create(maker: 'Bobbi Sox', title: 'Why Not?', kind: 'album', description: 'So tired')
+      test_case.votes.create
+
+    assert_difference("Vote.count", -1) do
+      get :downvote, {id: test_case.id}
+      assert_response :redirect
+    end
```
Comment (Hamled) [inline test/controllers/media_controller_test.rb:64]: We should also test that attempting to downvote when the medium has zero votes _doesn't_ change the number of votes in the DB.

File: test/controllers/media_controller_test.rb [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    assert_difference("Vote.count", 1) do
+      post :upvote, {id: media(:harold_and_maude).id}
+      assert_response :redirect
+    end
+  end
+
+  test 'Upvote Method Must Add Votes to Specific Medium' do
+    test_case = media(:flowers_in_the_attic)
+    test_case.votes.create
+    assert_equal( test_case.votes.count, 1)
```
Comment (Hamled) [inline test/controllers/media_controller_test.rb:54]: Because this test doesn't call any controller actions it's more appropriate for the model test file. Instead, we could write this as a controller test like so:

``` ruby
test 'Upvote Method Must Add Votes to Specific Medium' do
  assert_difference('media(:flowers_in_the_attic).votes.count', 1) do
    post :upvote, {id: media(:flowers_in_the_attic).id}
  end
end
```

File: test/models/medium_test.rb [hunk 7]
```diff
@@ -0,0 +1,37 @@
+require 'test_helper'
+
+class MediumTest < ActiveSupport::TestCase
+  # test "the truth" do
+  #   assert true
+  # end
+#validations
+  test 'A Medium Must Have a Title' do
+  	assert media(:flood).valid?
```
Comment (Hamled) [inline test/models/medium_test.rb:9]: In order for this to actually test that a medium must have a title, you should also:
1. Remove the title by setting it to nil
2. Assert that the medium object is no longer valid

File: test/models/medium_test.rb [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+# title cannot be just spaces
+# maker name cannot be just spaces
+# description cannot be just spaces
+
+
+
+
+  test 'A Medium Must Have a Maker' do
+  	assert_equal( media(:flood).maker, 'They Might Be Giants')
```
Comment (Hamled) [inline test/models/medium_test.rb:26]: This does not test that the medium must have a maker [to be valid], it only checks that the medium defined in the fixtures file has that particular maker.

File: test/models/medium_test.rb [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+  end
+
+  raise error if no maker
```
Comment (Hamled) [inline test/models/medium_test.rb:29]: This, as far as I can tell, is only going to cause errors when attempting to run your test suite.

File: test/models/medium_test.rb [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+  test 'A Medium Must Have a Description' do
+     assert_not_nil( media(:flood), :description)
```
Comment (Hamled) [inline test/models/medium_test.rb:32]: This does not test that a medium must have a description, it only checks that the fixture named `flood` is defined in the fixtures file.

File: app/views/media/_media_form.html.erb [hunk 11]
```diff
@@ -0,0 +1,16 @@
+<%= @error if @error %>
+<%= form_for @my_medium, method: @post_method, url: @post_path  do |f| %>
+
+	<%= f.hidden_field :kind, value: params[:kind] %>
```
Comment (Hamled) [inline app/views/media/_media_form.html.erb:4]: This should use `hidden_field_tag` instead of `f.hidden_field` because when we use `f.hidden_field` the resulting param is named `medium[kind]` and when we generate this form again (e.g. because there is a validation error when saving the new medium) we no longer have `param[:kind]`.

To reproduce this issue:
1. Click the link to create a new medium, then click "Submit Art" without filling in any fields.
2. You'll be shown the page again with the error message indicating you need to fill out all fields.
3. Fill out all fields and click "Submit Art" again.
4. Observe that you receive the error message again. This is because the re-rendered form no longer has `params[:kind]` so the hidden field is empty.

An alternative way to resolve this issue would be to change the controller and take out this hidden field entirely. See my next comment on the controller code for more details.

File: app/controllers/media_controller.rb [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+    @post_path = 'create'
+  end
+
+  def create
+    @params = params
+    @my_medium = Medium.new
+    @my_medium.title = params[:medium][:title]
+    @my_medium.maker = params[:medium][:maker]
+    @my_medium.description = params[:medium][:description]
+    @my_medium.kind = params[:medium][:kind]
```
Comment (Hamled) [inline app/controllers/media_controller.rb:36]: We should use `params[:kind]` here instead, and change the way we set `@post_path` in this action and the `new` action.

Instead of `@post_path = create_path` and `@post_path = 'create'` we should use `@post_path = create_path(kind: params[:kind])` in both locations.

File: app/controllers/media_controller.rb [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+    @my_medium.votes.create
+    redirect_to(index_path)
+  end
+
+  def downvote
+    @my_medium = Medium.find(params[:id])
+    if @my_medium.votes.count >= 1
+      @my_medium.votes.first.destroy
+    else
+      return "It's already at '0'. You can't downvote it further."
```
Comment (Hamled) [inline app/controllers/media_controller.rb:116]: This message needs to be communicated to the user, so it should go in a flash message rather than a `return` statement.

File: app/controllers/media_controller.rb [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+    # end  
+
+    @my_medium.title = params[:medium][:title]
+    @my_medium.maker = params[:medium][:maker]
+    @my_medium.description = params[:medium][:description]
+  
+    if @my_medium.save
+      redirect_to index_path
+    else
+      @error = "Did not save successfully. Try again!"
```
Comment (Hamled) [inline app/controllers/media_controller.rb:91]: This is good, but in order for us to see the error message set on this line, we'll need to also say `render :edit`. Check out how you've done it in the `create` action for an example of what we should be doing here as well.
